### PR TITLE
[RLlib] - Fix numerical overflow in gradient clipping for (many) large gradients

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

Large gradients and many of these could lead to numerical overflow when computing their l2-norm in `torch_utils.clip_gradients` (using the "global_norm"). This is counterproductive as a user wants to clip such gradients and instead runs into numerical overflow because of clipping gradients. 

This PR proposes small changes to turn `inf` and `neginf` values returned from norms to `10e8` and `-10e8`, respectively. This does not harm gradients themselves (if these for example were already `inf/neginf` b/c we clip gradients by multiplication and not overriding values).

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/utils/tests/test_torch_utils.py [hunk 1]
```diff
@@ -94,6 +95,29 @@ def test_copy_torch_tensors(self):
                 all(copied_tensor.detach().numpy() == tensor.detach().cpu().numpy())
             )
 
+    def test_large_gradients_clipping(self):
```
Comment (sven1977) [inline rllib/utils/tests/test_torch_utils.py:98]: Nice! Thanks for creating this important test case!

File: rllib/utils/torch_utils.py [hunk 2]
```diff
@@ -155,15 +155,26 @@ def clip_gradients(
...
-        )
+        ).nan_to_num(neginf=-10e8, posinf=10e8)
         if torch.logical_or(total_norm.isnan(), total_norm.isinf()):
             raise RuntimeError(
                 f"The total norm of order {norm_type} for gradients from "
                 "`parameters` is non-finite, so it cannot be clipped. "
             )
-        clip_coef = grad_clip / (total_norm + 1e-6)
+        clip_coef = grad_clip / torch.maximum(
+            torch.tensor(grad_clip), total_norm + 1e-6
```
Comment (sven1977) [inline rllib/utils/torch_utils.py:176]: Two questions:
* Would this `torch.tensor()` pose a danger when on another device? GPU?
* Can we add a comment  here (or enhance the one below) explaining why we compute the coeff like this? What are the expected final values of the coeff (between 0.0 and 1.0)?

**Review by sven1977 — APPROVED** [review]
Looks great! Thanks for fixing this very important piece of code @simonsays1980 . Just 2 nits and questions before we can merge.

File: rllib/utils/torch_utils.py (hunk 2 above)
Comment (simonsays1980) [reply rllib/utils/torch_utils.py:176]: Good catch! Yes, let's put the tensor on the `device` we have extracted before to be save. I will add then a note why we want the coefficient not larger than 1.0.

File: rllib/utils/torch_utils.py [hunk 3]
```diff
@@ -155,15 +155,29 @@ def clip_gradients(
...
+        ).nan_to_num(neginf=-10e8, posinf=10e8)
         if torch.logical_or(total_norm.isnan(), total_norm.isinf()):
             raise RuntimeError(
                 f"The total norm of order {norm_type} for gradients from "
                 "`parameters` is non-finite, so it cannot be clipped. "
             )
-        clip_coef = grad_clip / (total_norm + 1e-6)
+        # We do want the coefficient to be in between 0.0 and 1.0, therefore
+        # if the global_norm is smaller than the clip value, we use the clip value
+        # as normalization constant.
```
Comment (sven1977) [inline rllib/utils/torch_utils.py:177]: Perfect, thanks for making this comment much more clear!

**Comment by sven1977** [comment]
Looks great, let's merge once tests pass ...
